contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -4290,3 +4309,746 @@ sp_reset_connection_internal(PG_FUNCTION_ARGS)
+	delete_xml_handle_entry(doc_handle);
+
+	PG_RETURN_VOID();
+}
+
+/*
+ * Function to retrieve XML document from temporary table using document ID
+ */
+Datum
+tsql_openxml_get_xmldoc(PG_FUNCTION_ARGS)
+{
+	int32                  document_id = PG_GETARG_INT32(0);
+	Relation               relation;
+	ScanKeyData            skey[1];
+	TableScanDesc          scan;
+	HeapTuple              tuple;
+	bool                   found = false;
+	Datum                  result = (Datum) 0;
+	bool                   isnull = true;
+	EphemeralNamedRelation enr = NULL;
+	
+	/* Check if the temporary table exists */
+	if (xml_handle_temp_table_name != NULL)
+	{
+		enr = get_ENR(currentQueryEnv, xml_handle_temp_table_name, true);
+		if (enr)
+		{
+			relation = relation_open(enr->md.reliddesc, RowExclusiveLock);
+		}
+	}
+
+	if (!OidIsValid(enr->md.reliddesc))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+				 errmsg("XML document with ID %d not found", document_id)));
+	}
+
+	ScanKeyInit(&skey[0],
+				1,  /* Column number */
+				BTEqualStrategyNumber, F_INT4EQ,
+				Int32GetDatum(document_id));
+	
+	scan = table_beginscan_catalog(relation, 1, skey);
+	tuple = heap_getnext(scan, ForwardScanDirection);
+	
+	if (HeapTupleIsValid(tuple))
+	{
+		/* Get the XML document from column 6 (doc) */
+		result = heap_getattr(tuple, 6, RelationGetDescr(relation), &isnull);
+		
+		if (!isnull)
+		{
+			/* Make a copy of the value */
+			result = datumCopy(result, false, -1);
+			found = true;
+		}
+	}
+	
+	table_endscan(scan);
+	relation_close(relation, AccessShareLock);
+	
+	/* If we found the document, return it */
+	if (found)
+		PG_RETURN_DATUM(result);
+	
+	/* If we get here , we did not find the document */
+	ereport(ERROR,
+			(errcode(ERRCODE_UNDEFINED_OBJECT),
+			 errmsg("XML document with ID %d not found", document_id)));
+	
+	PG_RETURN_NULL();
+}
+
+/*
+ * tsql_openxml_get_colpattern - Generate XPath expressions for OPENXML columns
+ *
+ * This function generates the appropriate XPath expression for a column based on
+ * the OPENXML flag value. The flag determines whether to access XML data as
+ * attributes, elements, or either.
+ *   flag - Controls the XPath pattern generation:
+ *   0,1: Use attribute access (@colname)
+ *   2: Use element access (colname)
+ *   3: Try both element and attribute (colname|@colname)
+ */
+Datum
+tsql_openxml_get_colpattern(PG_FUNCTION_ARGS)
+{
+	char *xpath_expr;
+	int flag = PG_GETARG_INT32(1);
+	char *colname = text_to_cstring(PG_GETARG_TEXT_PP(0));
+	
+	switch (flag)
+	{
+		case 0:
+		case 1:
+			/* For flags 0 and 1, use @colname */
+			xpath_expr = psprintf("@%s", colname);
+			break;
+		case 2:
+			/* For flag 2, use colname */
+			xpath_expr = pstrdup(colname);
+			break;
+		case 3:
+			/* For flag 3, use colname|@colname */
+			xpath_expr = psprintf("%s|@%s", colname, colname);
+			break;
+		default:
+			/* Default to @colname for unknown flags */
+			xpath_expr = psprintf("@%s", colname);
```

Comment:
can colname contain '|' or '@' ? In that case, does xpath_expr need to be escaped?